fix(components): [tree-select] use up/down/enter keys select error

SpanManX opened the pull request:
close #19990
fix(components): [tree-select] fix background color to up/down keys, fix Enter selection bug.

**The bugs mentioned below can all be reproduced on the official website.**

**以下所提 BUG 都可在官网复现**

https://element-plus.org/zh-CN/component/tree-select.html

**Fix tree-select component bugs:**
1. No background color on element when pressing "Up/Down" keys.
2. Pressing Enter key does not select the data.

**This PR fixes issue #19990.**

**修改 tree-select 组件 BUG：**
1. 键盘按“上下键” 元素没有背景色。
2. 按 Enter 键没有选中数据。

**该 PR 修复了 issue #19990。**

### Original code test GIF images / 原始代码测试GIF图
![error](https://github.com/user-attachments/assets/a7f5519c-35bf-4919-bd85-1e6816d13371)

Clicking the triangle iconto expand ![QQ20250302-021446](https://github.com/user-attachments/assets/eeed64ad-53e3-4ba8-a91d-2f1d5cae7617) the tree and then pressing the "Up/Down" keys or Enter will trigger useKeydown.ts -> handleKeydown() instead of useSelect.ts -> navigateOptions().

The two different "console" logs printed in the console are to more clearly display and distinguish which function from which file was executed in the GIF.

点击三角形图标 ![QQ20250302-021446](https://github.com/user-attachments/assets/eeed64ad-53e3-4ba8-a91d-2f1d5cae7617) 展开树形后再按键盘“上下键或Enter键”，会执行 useKeydown.ts ->handleKeydown() ，而不是执行 useSelect.ts -> navigateOptions() 。

控制台打印的两个不同的 “console” 是为更直观的在 GIF 图上展示和区分执行了哪个文件下的函数。

### The test after changing the code / 修改代码后的测试
![success](https://github.com/user-attachments/assets/7f315599-ced5-4dd3-b11e-d695f1b3375a)

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

fix(components): [tree-select] fix background color to up/down keys, fix Enter selection bug.

SpanManX commented:
After modifying the code, I tested all the functionalities related to the select component, and no other functionalities were affected. If any issues are found that affect other functionalities, I can make further modifications.

SpanManX committed:
Merge branch 'dev' into dev

SpanManX commented:
@yujinpan 你好，请帮忙review一下这个PR。

yujinpan commented on packages/components/tree/src/tree-node.vue:
```diff
@@ -132,6 +133,10 @@ export default defineComponent({
+    selectFocus: {
+      type: Function,
+      default: null,
+    },
```
The new prop doesn't seem to belong in the tree, try using the `node-expand/node-collapse` events in the TreeSelect component instead.

SpanManX replied:
已删除 selectFocus，改为在 tree.ts onNodeExpand() 和 onNodeCollapse() 内调用 select.value?.focus()

yujinpan commented on packages/theme-chalk/src/tree.scss:
```diff
@@ -75,6 +75,9 @@
     }
+    &:has(.#{$state-prefix}hovering) {
+      background-color: getCssVar('tree-node-hover-bg-color');
+    }
```
Same as above.

SpanManX replied:
样式移动到了 tree-select.scss

yujinpan reviewed (dismissed):
Thanks for your contribution.

But there are still some issues to be fixed:

- Up/Down does not work in collapsed tree nodes (may have other effects when using `renderAfterExpand`)
- Left/Right does not expand/collapse tree nodes

Finally we need some tests to check the code.

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX commented:
> Thanks for your contribution.
> 
> But there are still some issues to be fixed:
> 
> * Up/Down does not work in collapsed tree nodes (may have other effects when using `renderAfterExpand`)
> * Left/Right does not expand/collapse tree nodes
> 
> Finally we need some tests to check the code.

我已经全部重写完毕，并且发现并修改了另外一个BUG。明天会更新上代码，并加上代码描述

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX commented:
> Thanks for your contribution.
> 
> But there are still some issues to be fixed:
> 
> * Up/Down does not work in collapsed tree nodes (may have other effects when using `renderAfterExpand`)
> * Left/Right does not expand/collapse tree nodes
> 
> Finally we need some tests to check the code.

@yujinpan 你好，已按照要求修改了代码，请帮忙review一下
Hello, I have modified the code as requested. Please help review it.

I modified the code implementation in a simpler way:

1. Switched to using the tree component's keyboard events.
2. Used the left and right arrow keys to expand and collapse the tree.
3. Pressing Enter checks whether a value can be selected. By default, pressing Enter on a parent node expands it, considering cases where checkStrictly and lazy properties are set to true.
4. Resolved the impact of lazy loading and renderAfterExpand on the background color display order when navigating with the up and down arrow keys.

我用更简单的方式修改了该代码实现的功能。

1. 改用 tree 组件的键盘事件。
2. 左右按键展开折叠 tree。
3. 回车判断是否可选中值。默认属性父级按下回车展开节点，并判断 checkStrictly 和 lazy 属性为 true 的情况。
4. 解决了 lazy 和 renderAfterExpand 加载数据对上下按键背景色显示顺序造成的影响。

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
fix(components): [tree-select] find the first enabled option

SpanManX committed:
Merge branch 'dev' into dev

SpanManX commented:
@yujinpan 你好，已按照要求修改了代码，请帮忙review一下

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX commented:
@yujinpan 你好，是否还需要修改？

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX committed:
Update useSelect.ts

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

fix(components): [tree-select] fix bg for up/down keys, Enter bug

SpanManX committed:
Update useSelect.ts

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX committed:
Merge branch 'element-plus:dev' into dev

Dsaquel commented on packages/components/tree-select/src/cache-options.ts:
```diff
@@ -48,6 +52,8 @@ export default defineComponent({
       { flush: 'post', immediate: true }
     )
 
+    select.states.isUseTreeKeydown = props.isUseTreeKeydown
```
because it will always be true we can set to `true`  i think

Dsaquel reviewed (commented):
I left some comment, but over all i think it's good :+1: .
As @yujinpan said, can you add a test case ?

SpanManX committed:
fix(components): [tree-select] fix bg for up/down keys, Enter bug

fix(components): [tree-select] fix background color to up/down keys, fix Enter selection bug.

SpanManX commented:
> I left some comment, but over all i think it's good 👍 . As @yujinpan said, can you add a test case ?

I’ve added test cases, but the trigger firing order in the test cases is different from that in the browser. Is there a demo I can refer to?

Dsaquel commented:
I also didn't find a way to test it. Input somehow can't yield active element. I'll see tomorrow how to do it.

Dsaquel commented:
With we can `input.trigger('keydown.down') `active element but somehow `.find` or `.querySelector` can't find `.hovering`. Even with [pr](https://github.com/element-plus/element-plus/pull/20499). Can't say if it's an issue or we are doing in the wrong way.
For now we can use `navigateOptions` to mock the keydown just like in [select.test.ts](https://github.com/element-plus/element-plus/blob/dev/packages/components/select/__tests__/select.test.ts#L784-L786).

Dsaquel commented on packages/components/select/src/useSelect.ts:
```diff
@@ -761,6 +778,8 @@ export const useSelect = (props: ISelectProps, emit: SelectEmits) => {
       }
       nextTick(() => scrollToOption(hoverOption.value))
+    } else if (states.isUseTreeKeydown) {
+      useTreeKeydown()
```
Should handle the update of `hoveringIndex` i think.

SpanManX replied:
In the case of using renderAfterExpand and lazy, the order of hoveringIndex becomes incorrect.

在使用 `renderAfterExpand`  和 `lazy` 的情况下，`hoveringIndex`  顺序是错乱的

SpanManX replied:
Okay, I'll look into whether there's a way to have the best of both worlds — adding test cases without changing the original logic.

Dsaquel commented:
Lgtm

SpanManX committed:
Merge branch 'element-plus:dev' into dev

SpanManX committed:
fix(components): [tree-select] Add data type

fix(components): [tree-select] Add data type

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
fix(components): [tree-select] change the keyboard methods

SpanManX commented:
> change the keyboard methods

I've already changed the keyboard methods   @Dsaquel  @warmthsea

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

Dsaquel commented:
Sorry, I just re-check [#15768](https://github.com/element-plus/element-plus/pull/15768) is just out of context in this pr i mixed up with something else.

My only blocking thing is the changes in the select folder.
I will checkout later if we can have a all changes just in tree-select folder.

SpanManX committed:
Merge branch 'element-plus:dev' into dev

Dsaquel committed:
chore: restore to dev

Dsaquel committed:
chore: set solution

Dsaquel commented on packages/components/tree/src/model/useKeydown.ts:
```diff
@@ -26,6 +26,32 @@ export function useKeydown({ el$ }: UseKeydownOption, store: Ref<TreeStore>) {
     })
   })
 
+  const handleNode = (currentItem: HTMLDivElement, code: string = '') => {
+    const key = currentItem.dataset.key
+    const node = key !== undefined ? store.value.getNode(key) : null
+    if (!node) {
+      // 处理节点未找到的情况，例如 key 属性不存在
+      console.error(
+        'The node was not found, possibly because the key property does not exist.'
+      )
+      return
+    }
+
+    // Check whether it is a leaf node or the last child node
+    if (
+      (!node.childNodes.length && !node.store.lazy) ||
+      (node.store.lazy && node.isLeaf) ||
+      (node.store.checkStrictly &&
+        ![EVENT_CODE.left, EVENT_CODE.right].includes(code))
+    ) {
+      currentItem.click()
+    } else if (!node.expanded) {
+      node.expand()
+    } else if (node.expanded) {
+      node.collapse()
+    }
+  }
```
I didn't find the difference with a regular click event. Can you explain why ?

Dsaquel commented on packages/components/tree-select/src/select.ts:
```diff
@@ -37,6 +38,28 @@ export const useSelect = (
   )
 
+  //TODO: remove this
+  onMounted(() => {
+    useEventListener(
+      () => select.value?.$el,
+      'keydown',
+      async (evt) => {
+        const { dropdownMenuVisible, hoverOption } = select.value!
+        if (EVENT_CODE.down === evt.key && dropdownMenuVisible) {
+          await nextTick()
+          // wait navigateOption to finish
+          setTimeout(() => {
+            // el-select-dropdown__item => el-tree-node__content => el-tree-node__content
+            hoverOption?.$el?.parentNode?.parentNode?.focus()
+          })
+        }
+      },
+      {
+        capture: true,
+      }
+    )
+  })
```
@SpanManX I took some code of your previous solution and setted here.
From now this solution will fix the down event but can't resolve the up event. 
The best we can have is [handleKeydown](https://github.com/element-plus/element-plus/blob/dev/packages/components/tree/src/model/useKeydown.ts#L29-L109) for a good solution.
But until we refact it, i think we can merge this.

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

SpanManX committed:
Merge branch 'dev' into dev

Dsaquel commented:
@SpanManX I see you have got the `mousedown.up` working— so cool!

I will review a bit later.

Dsaquel commented:
Sometimes the keydown up/down is blocked after expanding some nodes:
![Peek 2025-07-28 00-41](https://github.com/user-attachments/assets/d680f68c-a13d-48c0-a02a-01fee2ef0971)

SpanManX committed:
Merge branch 'dev' into dev

Dsaquel commented on packages/components/tree-select/src/select.ts:
```diff
@@ -37,6 +38,49 @@ export const useSelect = (
   )
 
+  const focusLastNode = (listNode) => {
+    if (listNode.at(-1).expanded && listNode.at(-1).childNodes.at(-1)) {
+      focusLastNode([listNode.at(-1).childNodes.at(-1)])
```
**nit**:

I think we can simplify this.

Dsaquel commented on packages/components/tree-select/src/select.ts:
```diff
@@ -37,6 +38,49 @@ export const useSelect = (
+      const el = tree.value.el$?.querySelector(
+        `[data-key="${listNode.at(-1).key}"]`
+      )
+      el.focus()
```
```suggestion
      el?.focus()
```

Dsaquel commented on packages/components/tree-select/src/tree.ts:
```diff
@@ -179,7 +179,8 @@ export const useTree = (
         e.proxy.handleExpandIconClick()
       }
-      select.value?.focus()
+      // TODO: Using this method or not makes no difference
+      // select.value?.focus()
```
Let's keep it for now.

SpanManX replied:
Using this method may result in situations like this [ref](https://github.com/element-plus/element-plus/pull/20008#issuecomment-3051965051)

![464114002-7980d6ad-3075-4b55-b931-95dbfe42eb96](https://github.com/user-attachments/assets/65f5fa4d-a4c2-4d52-bca6-8ffd363656dc)

Dsaquel replied:
Okay didn't notice that, we can remove both line then 👍 .

Dsaquel reviewed (commented):
I forgot we merged the fixed old pr, added just some nits.

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
Update packages/components/tree-select/src/select.ts

Co-authored-by: Noblet Ouways <91417411+Dsaquel@users.noreply.github.com>

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

Dsaquel reviewed (approved):
Thanks!

SpanManX committed:
fix(components): [tree-select] use up/down/enter keys select error

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
fix(components): [tree-select] Fix code conflicts

SpanManX committed:
Merge branch 'dev' into dev

SpanManX committed:
Merge branch 'dev' into dev

tolking committed:
Merge branch 'dev' into dev

tolking reviewed (commented):
LGTM.

It would be best if some details could be further optimized

SpanManX committed:
Apply suggestions from code review

Co-authored-by: qiang <qw13131wang@gmail.com>

SpanManX committed:
fix(components): [tree-select] Apply suggestions from code review

SpanManX committed:
fix(components): [tree-select] Apply suggestions from code review

Co-authored-by: qiang <qw13131wang@gmail.com>

SpanManX commented:
@yujinpan  你好，请帮忙review一下这个PR。
